src/WebJobs.Script/Diagnostics/FileTraceWriter.cs

```diff
@@ -39,90 +40,141 @@ public FileTraceWriter(string logFilePath, TraceLevel level, LogType logType) :
             _instanceId = GetInstanceId();
             _logType = logType;
             _logDirectory = new DirectoryInfo(logFilePath);
-
-            // start a timer to flush accumulated logs in batches
-            _flushTimer = new Timer
-            {
-                AutoReset = true,
-                Interval = LogFlushIntervalMs
-            };
-            _flushTimer.Elapsed += OnFlushLogs;
-            _flushTimer.Start();
         }
 
-        public override void Flush()
+        /// <summary>
+        /// Maintains a collection of objects used for locking, scoped to the file path.
+        /// The objects are not currently removed, but the list is pretty static and shouldn't be a concern.
+        /// These cannot be removed on disposal. If cleanup functionality is ever required, we can address this with a scheduled cleanup or something along those lines.
+        /// </summary>
+        /// <returns>A lock object scoped to the current <see cref="FileTraceWriter"/> log file path.</returns>
+        private object GetFlushLock() => _flushLocks.GetOrAdd(_logFilePath, new object());
```

Comment:
Locks are now scoped to a given path. This change means that function and host trace writers are no longer attempting to acquire the same lock.